micrometer-core/src/main/java/io/micrometer/core/instrument/binder/httpcomponents/AbstractPoolingHttpClientConnectionManagerMetricsBinder.java

```diff
@@ -0,0 +1,103 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder.httpcomponents;
+
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.lang.NonNull;
+import org.apache.http.conn.routing.HttpRoute;
+import org.apache.http.pool.ConnPoolControl;
+import org.apache.http.pool.PoolStats;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Base class which collects total metrics including {@link HttpRoute} state from {@link ConnPoolControl}, for example {@link org.apache.http.impl.conn.PoolingHttpClientConnectionManager}
+ * for synchronous HTTP clients or {@link org.apache.http.impl.nio.conn.PoolingNHttpClientConnectionManager} for asynchronous HTTP clients.
+ * <p>
+ * It monitors the overall and per route connection pool state.
+ *
+ * @author Mikhail Yakimchenko (mikhailyakimchenko@gmail.com)
+ * @since 1.5.0
+ */
+abstract class AbstractPoolingHttpClientConnectionManagerMetricsBinder<T extends ConnPoolControl<HttpRoute>> extends
+        ConnPoolControlMetricsBinder<HttpRoute> {
+
+    @SuppressWarnings("unchecked")
+    private final T connectionManager = (T) connPoolControl;
+    private final ConcurrentMap<String, PoolStats> poolStatsByHost = new ConcurrentHashMap<>();
+
+    public AbstractPoolingHttpClientConnectionManagerMetricsBinder(T connectionManager, String name, String... tags) {
+        this(connectionManager, name, Tags.of(tags));
+    }
+
+    public AbstractPoolingHttpClientConnectionManagerMetricsBinder(T connectionManager, String name, Iterable<Tag> tags) {
+        super(connectionManager, name, tags);
+    }
+
+    protected abstract Collection<HttpRoute> getRoutes(T connectionManager);
+
+    @Override
+    public void bindTo(@NonNull MeterRegistry registry) {
+        super.bindTo(registry);
+        updateRoutesMetrics(registry);
+    }
+
+    /**
+     * Adds metrics for newly created {@link HttpRoute} to the given meter registry.
+     */
+    public void updateRoutesMetrics(MeterRegistry registry) {
+        for (HttpRoute route : getRoutes(connectionManager)) {
+            String hostName = route.getTargetHost().getHostName();
+            if (!poolStatsByHost.containsKey(hostName)) {
+                PoolStats stats = connectionManager.getStats(route);
+                poolStatsByHost.put(hostName, stats);
+                registerRouteMetrics(hostName, stats, registry);
+            }
+        }
+    }
+
+    private void registerRouteMetrics(String host, PoolStats poolStats, MeterRegistry registry) {
+        Gauge.builder("httpcomponents.httpclient.pool.route.max",
+                poolStats,
+                PoolStats::getMax)
```

Comment:
I know this is year old, but i've come across this trying to implement it in my app.
The information below applies at least to `PoolingHttpClientConnectionManager`.

You can't use/reuse the PoolStats object here, because the object is created when `connectionManager.getStats(route);` is called and the primitive values are filled into it. It is an immutable object, so your statistics will permanently display the state they were in the moment registerRouteMetrics was first called for the host.

I've worked around this by passing the HttpRoute object into the registerRouteMetrics and calling getStats in the Gauge lambda. This seems to work correctly even if the connection pool completely empties and the Route is no longer returned in getStats. However it will need further testing to see if it works correctly in all cases.